x/ibc/testing/coordinator.go

```diff
@@ -0,0 +1,178 @@
+	globalStartTime = time.Date(2020, 1, 2, 0, 0, 0, 0, time.UTC)
+	timeIncrement   = time.Second * 5
+)
+
+// Coordinator is a testing struct which contains N TestChain's. It handles keeping all chains
+// in sync with regards to time.
+type Coordinator struct {
+	Chains []*TestChain
+}
+
+// NewCoordinator initializes Coordinator with N TestChain's
+func NewCoordinator(t *testing.T, n uint64) *Coordinator {
+	chains := make([]*TestChain, n)
+
+	for i := range chains {
+		chains[i] = NewTestChain(t, ClientIDPrefix+string(i))
+	}
+	return &Coordinator{
+		Chains: chains,
+	}
+}
+
+// IncrementTime iterates through all the TestChain's and increments their current header time
+// by 5 seconds.
+//
+// CONTRACT: this function must be called after every commit on any TestChain.
+func (coord *Coordinator) IncrementTime() {
+	for _, chain := range coord.Chains {
+		chain.CurrentHeader = abci.Header{
+			Height: chain.CurrentHeader.Height,
+			Time:   chain.CurrentHeader.Time.Add((timeIncrement)),
+		}
+	}
+}
+
+// CommitBlock commits a block on the provided indexes and then increments the global time.
+//
+// CONTRACT: the passed in list of indexes must not contain duplicates
+func (coord *Coordinator) CommitBlock(chains ...uint64) {
+	for _, index := range chains {
+		chain := coord.Chains[index]
+		chain.App.Commit()
+		nextBlock(chain)
+	}
+	coord.IncrementTime()
+}
+
+// CommitNBlocks commits n blocks to state and updates the block height by 1 for each commit.
+func (coord *Coordinator) CommitNBlocks(index, n uint64) {
+	chain := coord.Chains[index]
+
+	for i := uint64(0); i < n; i++ {
+		chain.App.BeginBlock(abci.RequestBeginBlock{Header: chain.CurrentHeader})
+		chain.App.Commit()
+		nextBlock(chain)
+		coord.IncrementTime()
+	}
+}
+
+// CreateClient creates a counterparty client on the source chain.
+func (coord *Coordinator) CreateClient(source, counterparty uint64) {
+	coord.CommitBlock(source, counterparty)
+	coord.Chains[source].CreateClient(coord.Chains[counterparty])
+	coord.IncrementTime()
+}
+
+// UpdateClient updates a counterparty client on the source chain.
+func (coord *Coordinator) UpdateClient(source, counterparty uint64) {
+	coord.CommitBlock(source, counterparty)
+	coord.Chains[source].UpdateClient(coord.Chains[counterparty])
+	coord.IncrementTime()
+}
+
+// CreateConnection constructs and executes connection handshake messages in order to create
+// channels on source and counterparty chains with the passed in Connection State. The
+// connectionID's of the source and counterparty are returned.
+//
+// NOTE: The counterparty testing connection will be created even if it is not created in the
+// application state.
+func (coord *Coordinator) CreateConnection(
+	sourceIndex, counterpartyIndex uint64,
+	state connectiontypes.State,
+) (sourceConnection, counterpartyConnection string) {
+	source := coord.Chains[sourceIndex]
+	counterparty := coord.Chains[counterpartyIndex]
+
+	if state == connectiontypes.UNINITIALIZED {
+		return
+	}
```

Comment:
imo we should remove these checks so that if the handshake fails we know that something is misconfigured or we performed the wrong steps during the handshake

Reply:
So my main motivation for adding these was allowing the caller to initialize a conn/channel with an intermediate state while keeping the code DRY. I guess since using intermediate states is less likely we could push that onto the developer or add funcs that only execute the intermediate state despite the redundant code

Reply:
Yea makes sense to have these returns so we can return partially created connections to our test functions